src/Parser/MarksmanshipHunter/Modules/FocusChart/Focus.js

```diff
@@ -0,0 +1,353 @@
+import SPELLS from 'common/SPELLS';
+
+import specialEventIndicators from 'Main/Chartist/specialEventIndicators';
+
+import { formatDuration } from 'common/format';
+
+import 'Main/Mana.css';
+
+import FocusComponent from './FocusComponent';
+import './Focus.css';
+
+
+const passiveWasteThresholdPercentage = .03; // (wasted passive focus generated) / (total passive focus generated), anything higher will trigger "CAN BE IMPROVED"
+//TODO: get a "real" number approved by a MMS expert
+
+class Focus extends React.PureComponent {
+  static propTypes = {
+    reportCode: PropTypes.string.isRequired,
+    actorId: PropTypes.number.isRequired,
+    start: PropTypes.number.isRequired,
+    end: PropTypes.number.isRequired,
+    playerHaste: PropTypes.number.isRequired,
+    focusMax: PropTypes.number,
+    passiveWaste: PropTypes.array,
+    pWTracker : PropTypes.number,
+    activeWaste: PropTypes.array,
+    aWTracker: PropTypes.number,
+  };
+
+  constructor() {
+    super();
+
+    this.state = {
+      focusData: null,
+      bossHealth: null,
+    };
+	
+  }
+
+  componentWillMount() {
+    this.load(this.props.reportCode, this.props.actorId, this.props.start, this.props.end);      
+    }
+
+  componentWillReceiveProps(newProps) {
+    if (newProps.reportCode !== this.props.reportCode || this.props.focusMax !== this.props.focusMax || this.props.pWTracker !== this.props.pWTracker || this.props.aWTracker !== this.props.aWTracker || newProps.actorId !== this.props.actorId || newProps.start !== this.props.start || newProps.end !== this.props.end) {
+      this.load(newProps.reportCode, newProps.actorId, newProps.start, newProps.end);
+    }
+  }
+  load(reportCode, actorId, start, end) {
+
+ 
+    const focusPromise = fetch(makeWclUrl(`report/tables/resources/${reportCode}`, {
+      start,
+      end,
+      sourceid: actorId,
+      abilityid: 102,
+    })) 
+      .then(response => response.json())
+      .then((json) => {
+        if (json.status === 400 || json.status === 401) {
+          throw json.error;
+        } else {
+          this.setState({
+            focusData: json,
+          });
+        }
+      });
+    const bossHealthPromise = fetch(makeWclUrl(`report/tables/resources/${reportCode}`, {
+      start,
+      end,
+      sourceclass: 'Boss',
+      hostility: 1,
+      abilityid: 1000,
+    }))
+      .then(response => response.json())
+      .then((json) => {
+        if (json.status === 400 || json.status === 401) {
+          throw json.error;
+        } else {
+          this.setState({
+            bossHealth: json,
+          });
+        }
+      });
+
+    return Promise.all([focusPromise, bossHealthPromise]);
+  }
+
+  render() {
+        if (!this.state.focusData || !this.state.bossHealth || !this.props.passiveWaste) {
+      return (
+        <div>
+          Loading...
+        </div>
+      );
+    }
+
+	const actorId = this.props.actorId;
+  	const maxFocus = this.props.focusMax;
```

Comment:
This block of code looks pretty grouped, could it be extracted to its own method? Would make it easier to understand (and later maintain).

I'm thinking without fully understanding what's happening here that this might be better suited for a specific module.